pkg/tokenreviews/tokenvalidation/tokenauthenticator.go

```diff
@@ -1,84 +1,191 @@
+	userv1client "github.com/openshift/client-go/user/clientset/versioned/typed/user/v1"
+	bootstrap "github.com/openshift/library-go/pkg/authentication/bootstrapauthenticator"
+
+	tokenvalidators "github.com/openshift/oauth-apiserver/pkg/tokenreviews/tokenvalidation/validators"
+)
+
+const (
+	clusterAdminGroup       = "system:cluster-admins"
+	authenticatedOAuthGroup = "system:authenticated:oauth"
+
+	sha256Prefix = "sha256~"
 )
 
 var errLookup = errors.New("token lookup failed")
 
-type tokenAuthenticator struct {
-	tokens       oauthclient.OAuthAccessTokenInterface
-	users        userclient.UserInterface
-	groupMapper  UserToGroupMapper
-	validators   OAuthTokenValidator
-	implicitAuds kauthenticator.Audiences
+var _ authenticator.Request = &TokenAuthenticator{}
+
+type TokenAuthenticator struct {
+	accessTokenClient   oauthv1client.OAuthAccessTokenInterface
+	userClient          userv1client.UserInterface
+	bootstrapUserGetter bootstrap.BootstrapUserDataGetter
+	groupMapper         UserToGroupMapper
+
+	validators tokenvalidators.OAuthTokenValidator
 }
 
-func NewTokenAuthenticator(tokens oauthclient.OAuthAccessTokenInterface, users userclient.UserInterface, groupMapper UserToGroupMapper, implicitAuds kauthenticator.Audiences, validators ...OAuthTokenValidator) kauthenticator.Token {
-	return &tokenAuthenticator{
-		tokens:       tokens,
-		users:        users,
-		groupMapper:  groupMapper,
-		validators:   OAuthTokenValidators(validators),
-		implicitAuds: implicitAuds,
+func NewTokenAuthenticator(
+	accessTokenClient oauthv1client.OAuthAccessTokenInterface,
+	bootstrapUserGetter bootstrap.BootstrapUserDataGetter,
+	userClient userv1client.UserInterface,
+	groupMapper UserToGroupMapper,
+	validators ...tokenvalidators.OAuthTokenValidator,
+) *TokenAuthenticator {
+	return &TokenAuthenticator{
+		accessTokenClient:   accessTokenClient,
+		userClient:          userClient,
+		bootstrapUserGetter: bootstrapUserGetter,
+		groupMapper:         groupMapper,
+
+		validators: tokenvalidators.OAuthTokenValidators(validators),
 	}
 }
 
-func (a *tokenAuthenticator) AuthenticateToken(ctx context.Context, name string) (*kauthenticator.Response, bool, error) {
-	token, err := a.tokens.Get(context.TODO(), name, metav1.GetOptions{})
-	if err != nil {
-		return nil, false, errLookup // mask the error so we do not leak token data in logs
+func (h *TokenAuthenticator) AuthenticateRequest(req *http.Request) (*authenticator.Response, bool, error) {
+	authorizationHeader := strings.TrimSpace(req.Header.Get("Authorization"))
+	if len(authorizationHeader) == 0 {
+		return nil, false, nil
 	}
 
-	user, err := a.users.Get(context.TODO(), token.UserName, metav1.GetOptions{})
-	if err != nil {
-		return nil, false, err
+	authzHeaderSplit := strings.Split(authorizationHeader, " ")
+	if len(authzHeaderSplit) < 2 || strings.ToLower(authzHeaderSplit[0]) != "bearer" {
+		return nil, false, nil
```

Comment:
why all these changes here? I thought it's a direct copy from apiserver-library-go?!

Reply:
the authenticator for the upstream tokenreview registry has a slightly different interface (`"k8s.io/apiserver/pkg/authentication/authenticator"`) with slightly different expectations to the authenticator we had before (`"k8s.io/apiserver/pkg/authentication/authenticator"`).

The changes that follow only accommodate and leverage this change by unifying the two OpenShift-specific token authenticators we used to have into a single user-retrieving mechanism, duality of which has caused us such headaches in the past.